pkg/migration-tool/writer/shards.go

```diff
@@ -0,0 +1,186 @@
+// to handling and management of shards.
+type shardsSet struct {
+	num         int
+	set         []*shard
+	cancelFuncs []context.CancelFunc
+
+	errChan chan error
+}
+
+// newShardsSet creates a shards-set and initializes it. It creates independent clients for each shard,
+// contexts and starts the shards. The shards listens to the writer routine, which is responsible for
+// feeding the shards with data blocks for faster (due to sharded data) flushing.
+func newShardsSet(writerCtx context.Context, writeURL string, numShards int) (*shardsSet, error) {
+	var (
+		set         = make([]*shard, numShards)
+		cancelFuncs = make([]context.CancelFunc, numShards)
+	)
+	ss := &shardsSet{
+		num:     numShards,
+		errChan: make(chan error, numShards),
+	}
+	for i := 0; i < numShards; i++ {
+		client, err := utils.NewClient(fmt.Sprintf("writer-shard-%d", i), writeURL, utils.Write, model.Duration(defaultWriteTimeout))
+		if err != nil {
+			return nil, fmt.Errorf("creating write-shard-client-%d: %w", i, err)
+		}
+		ctx, cancelFunc := context.WithCancel(writerCtx)
+		shard := &shard{
+			ctx:          ctx,
+			client:       client,
+			queue:        make(chan *[]prompb.TimeSeries),
+			copyShardSet: ss,
+		}
+		cancelFuncs[i] = cancelFunc
+		set[i] = shard
+	}
+	ss.set = set
+	ss.cancelFuncs = cancelFuncs
+	// Run the shards.
+	for i := 0; i < ss.num; i++ {
+		go ss.set[i].run(i)
+	}
+	return ss, nil
+}
+
+// scheduleTS batches the time-series based on the available shards. These batches are then
+// fed to the shards in a single go which then consume and push to the respective clients.
+func (s *shardsSet) scheduleTS(ts []prompb.TimeSeries) (numSigExpected int) {
+	var batches = make([][]prompb.TimeSeries, s.num)
+	for _, series := range ts {
+		hash := utils.HashLabels(labelsSlicetoLabels(series.GetLabels()))
+		batchIndex := hash % uint64(s.num)
+		batches[batchIndex] = append(batches[batchIndex], series)
+	}
+	// Feed to the shards.
+	for shardIndex := 0; shardIndex < s.num; shardIndex++ {
+		batchIndex := shardIndex
+		if len(batches[batchIndex]) == 0 {
+			// We do not want "wait state" to happen on the shards.
+			continue
+		}
+		numSigExpected++
+		s.set[shardIndex].queue <- &batches[batchIndex]
+	}
+	return
+}
+
+// shard is an atomic unit of writing. Each shard accepts a sharded time-series set and
+// pushes the respective time-series via a POST request. More shards will lead to bigger
+// throughput if the target system can handle it.
+type shard struct {
+	ctx          context.Context
+	client       *utils.Client
+	queue        chan *[]prompb.TimeSeries
+	copyShardSet *shardsSet
+}
+
+// run runs the shard.
+func (s *shard) run(shardIndex int) {
+	defer func() {
+		log.Info("msg", fmt.Sprintf("shard-%d is inactive", shardIndex))
+	}()
+	log.Info("msg", fmt.Sprintf("shard-%d is active", shardIndex))
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case refTs, ok := <-s.queue:
+			if !ok {
+				return
+			}
+			if len(*refTs) != 0 {
```

Comment:
I'd actually consider throwing a panic if len ==0, that represents and irrecoverable logic error.

Reply:
I feel like panic is something severe and may silently stall the system given the future plans of streaming. An error logging with details would be better without exiting the shards.